### DISCUSSION TITLE
Is there a way to calculate overlap percentage/score from a perspective of smaller chemical structure?

### DISCUSSION THREAD

**Discussion by jrkarki** [rdkit/rdkit#5983] [Q&A]
Hi everyone,

I'm trying to identify individual components of a hetero-molecule. For example, for A = B + C + D, taking A as the reference, I want to find B from B_list = [B1,B2,B3,...,Bn]. Identifying B works using A.HasSubstructMatch(B1/B2..Bn) from rdkit, however it is true only if there is 100% match. Can I identify B from B_list which has a match of say 99% with A.

I can't seem to find a way to calculate scores for items in B_list against A. The fingerprint (FP) scores are not helping because I want to get scores purely from perspective of B. The FP metrics work both ways!!

I tried using FP metric i.e., DataStructs.FingerprintSimilarity(A,B) but this doesn't seem to work as I expected. I generated FP for each B in B_list against A and used a cut-off to filter the list. However, my B (see below) with best FP metric doesn't seem to be a sub-structure of A. It is something else.

For your ref: A = 'COC1=CC(C2=CN(C)C(=O)C3=CN=CC=C23)=CC(OC)=C1CN1CCN(CCOCCOCC(=O)NCC2=CC=C(S(=O)(=O)NC3=CC=CC4=C3[NH]C=C4Cl)C=C2)CC1'

B_with_best_FP = 'CCOC1=CC(C(C)(C)C)=CC=C1C1=N[C@@](C)(C2=CC=C(Cl)C=C2)[C@@](C)(C2=CC=C(Cl)C=C2)N1C(=O)N1CCN(CCCS(C)(=O)=O)CC1'

My_ideal_B = B from B_list which has maximum sub-structure match with A with a score. B_with_best_FP isn't the one.

B_list = https://drive.google.com/file/d/1BjcY8BLJOE98-OQT_zRSoRr7tWcO72Nt/view?usp=sharing

Any suggestions?

Thanks,
Reagon

**Comment by greglandrum** [comment]
@jrkarki you haven't said which fingerprint you used, so I can't comment specifically on that.
I think you need to use a fingerprint which has the "substructure relationship": i.e. if a molecule B is a substructure of molecule A then every bit set in the fingerprint of molecules B will also be set in the fingerprint of molecule A.
I would recommend that you use the RDKit's pattern fingerprint for this. You can generate that with:
```
Chem.PatternFingerprint(mol)
```
Here's a demonstration of that:
```
In [7]: fp_a = Chem.PatternFingerprint(mol_a)

In [8]: fp_b = Chem.PatternFingerprint(mol_b)

In [9]: DataStructs.AllProbeBitsMatch(fp_b,fp_a)
Out[9]: True

In [10]: DataStructs.AllProbeBitsMatch(fp_a,fp_b)
Out[10]: False
```
One straightforward tool for calculating the kind of symmetry I think you want is the asymmetric similarity, which looks at the number of bits in common divided by the number of bits in the fingerprint with less bits set:
```
In [11]: DataStructs.AsymmetricSimilarity(fp_a,fp_b)
Out[11]: 1.0
```

**Reply by jrkarki** [reply]
Thanks a lot Greg. I will follow your suggestions and will let you know soon. Btw, I had tried topological and MACCS Keys FPs.

**Comment by jrkarki** [comment]
Hi Greg, 

Thank you once again. I tried a no. of solutions but Maximum Common Sub-structure (MCS) seemed to work best in my case. I performed a pairwise MCS using rdFMCS.FindMCS(mols) and filtered the results with highest molecular weight. This would eventually be my sub-structure I was looking for. However, double bonds are lost in the SMILES of the MCS. I tried to kekulize the SMILES after getting rid of the symbol ":" that appear in the MCS but no luck. Any suggestions to get the double bonds back? 

mcs = 'O=S(=O)(NC1:C:C:C:C2:C1:N:C:C2Cl)C1:C:C:CC:C:1'

full_mol = 'COC1=CC(C2=CN(C)C(=O)C3=CN=CC=C23)=CC(OC)=C1CN1CCN(CCOCCOCC(=O)NCC2=CC=C(S(=O)(=O)NC3=CC=CC4=C3[NH]C=C4Cl)C=C2)CC1'

![mcs1](https://user-images.githubusercontent.com/25939327/215439905-bc475eea-a460-4b66-85db-2ab0175621c9.png)
![mcs2](https://user-images.githubusercontent.com/25939327/215439954-1c4c920d-53bf-433e-84d6-949eaddaeabc.png)
![Full_mol](https://user-images.githubusercontent.com/25939327/215440513-481e2bff-31cf-453b-be12-a3aa94cb8e66.png)

The figures will make it more clearer.
